dask/tests/test_base.py

```diff
@@ -572,45 +622,95 @@ def __dask_postcompute__(self):
         return tuple, ()
 
     def __dask_postpersist__(self):
-        return Tuple, (self._keys,)
+        def rebuild(dsk, name=None):
+            keys = self._keys
+            if name is not None:
+                keys = [replace_name_in_key(key, name) for key in keys]
+            return Tuple(dsk, keys)
+
+        return rebuild, ()
 
 
 def test_custom_collection():
-    dsk = {"a": 1, "b": 2}
-    dsk2 = {"c": (add, "a", "b"), "d": (add, "c", 1)}
+    # Arbitrary hashables
+    h1 = object()
+    h2 = object()
+
+    # Collections with 2+ keys must have all keys in the format of tuples where the
+    # first element is the same string, referred to as collection name, and the rest are
+    # arbitrary hashables
+    dsk = {("x", h1): 1, ("x", h2): 2}
+    dsk2 = {("y", h1): (add, ("x", h1), ("x", h2)), ("y", h2): (add, ("y", h1), 1)}
     dsk2.update(dsk)
-    dsk3 = {"e": (add, "a", 4), "f": (inc, "e")}
-    dsk3.update(dsk)
 
-    x = Tuple(dsk, ["a", "b"])
-    y = Tuple(dsk2, ["c", "d"])
-    z = Tuple(dsk3, ["e", "f"])
+    # If and only if there is only one top-level key, it can be just a string
+    dsk3 = {"z": (add, ("y", h1), ("y", h2))}
+    dsk3.update(dsk2)
+
+    w = Tuple({}, [])  # A collection can have no keys at all
+    x = Tuple(dsk, [("x", h1), ("x", h2)])
+    y = Tuple(dsk2, [("y", h1), ("y", h2)])
+    z = Tuple(dsk3, ["z"])
 
     # __slots__ defined on base mixin class propagates
     with pytest.raises(AttributeError):
         x.foo = 1
 
     # is_dask_collection
+    assert is_dask_collection(w)
     assert is_dask_collection(x)
+    assert is_dask_collection(y)
+    assert is_dask_collection(z)
 
     # tokenize
+    assert tokenize(w) == tokenize(w)
     assert tokenize(x) == tokenize(x)
-    assert tokenize(x) != tokenize(y)
+    assert tokenize(y) == tokenize(y)
+    assert tokenize(z) == tokenize(z)
+    assert len({tokenize(coll) for coll in (w, x, y, z)}) == 4
+
+    # get_collection_name
+    with pytest.raises(KeyError):
+        get_collection_name(w)
+    assert get_collection_name(x) == "x"
+    assert get_collection_name(y) == "y"
+    assert get_collection_name(z) == "z"
 
     # compute
+    assert w.compute() == ()
     assert x.compute() == (1, 2)
-    assert dask.compute(x, [y, z]) == ((1, 2), [(3, 4), (5, 6)])
-    t = x + y + z
-    assert t.compute() == (1, 2, 3, 4, 5, 6)
+    assert y.compute() == (3, 4)
+    assert z.compute() == (7,)
+    assert dask.compute(w, [{"x": x}, y, z]) == ((), [{"x": (1, 2)}, (3, 4), (7,)])
+    t = w + x + y + z
+    assert t.compute() == (1, 2, 3, 4, 7)
 
     # persist
     t2 = t.persist()
     assert isinstance(t2, Tuple)
-    assert t2._dask == dict(zip("abcdef", range(1, 7)))
-    assert t2.compute() == (1, 2, 3, 4, 5, 6)
-    x2, y2, z2 = dask.persist(x, y, z)
+    assert t2._keys == t._keys
```

Comment:
That was because I had to complicate how ``Tuple.__add__`` works to make it compliant to the new protocol spec. I now partially reintegrated the test.